### PR TITLE
JDK-8267916: Adopt cast notation for CompilerThread conversions

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring around `as_CompilerThread`. A new file, `compilerThread.inline.hpp`, is created to get around the circular dependency.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267916](https://bugs.openjdk.java.net/browse/JDK-8267916): Adopt cast notation for CompilerThread conversions

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4240/head:pull/4240` \
`$ git checkout pull/4240`

Update a local copy of the PR: \
`$ git checkout pull/4240` \
`$ git pull https://git.openjdk.java.net/jdk pull/4240/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4240`

View PR using the GUI difftool: \
`$ git pr show -t 4240`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4240.diff">https://git.openjdk.java.net/jdk/pull/4240.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Sorry I'm not seeing the need for a new compilerThread.inline.hpp. That isn't making things consistent AFAICS - if we haven't factored out CompilerThread into its own header we shouldn't have compilerThread.inline.hpp.

David

**Comment by albertnetymk** [comment]
> That isn't making things consistent

It's the renaming and move to `thread.inline.hpp` (two points covered in the ticket description) that is supposed to make it more consistent. The creation of `compilerThread.inline.hpp` is just to work around the circular dependency, `CompilerThread::current()` in `compilerThread.hpp` calls `JavaThread::as_CompilerThread` (in `thread.inline.hpp` after the move).

**Comment by dholmes-ora** [comment]
Sorry I looked at this too late at night:

>  if we haven't factored out CompilerThread into its own header

but we do of course have that.

File: src/hotspot/share/ci/ciEnv.hpp [hunk 1]
```diff
@@ -31,7 +31,7 @@
 #include "code/debugInfoRec.hpp"
 #include "code/dependencies.hpp"
 #include "code/exceptionHandlerTable.hpp"
-#include "compiler/compilerThread.hpp"
+#include "compiler/compilerThread.inline.hpp"
```
Comment (kimbarrett) [inline src/hotspot/share/ci/ciEnv.hpp:34]: Including .inline.hpp by a .hpp is not allowed.

File: src/hotspot/share/compiler/compilerThread.inline.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_COMPILER_COMPILERTHREAD_INLINE_HPP
+#define SHARE_COMPILER_COMPILERTHREAD_INLINE_HPP
+
+#include "runtime/thread.inline.hpp"
+#include "compiler/compilerThread.hpp"
```
Comment (kimbarrett) [inline src/hotspot/share/compiler/compilerThread.inline.hpp:29]: This include order is inconsistent with JDK-8267464.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by albertnetymk** [comment]
I encountered some circular dependencies after moving the definition to `thread.inline.hpp`, so I replaced it with a static `cast` method, as Kim instructed from an offline discussion. Note now there's a discrepancy: `CompilerThread::cast(t)` vs `t->as_Worker_thread()` (or `t->as_Java_thread()`). I can do the same for other `as_*_thread` methods as well (in this PR or another one) if people feel like this approach.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by kimbarrett** [comment]
> I encountered some circular dependencies after moving the definition to `thread.inline.hpp`, so I replaced it with a static `cast` method, as Kim instructed from an offline discussion. Note now there's a discrepancy: `CompilerThread::cast(t)` vs `t->as_Worker_thread()` (or `t->as_Java_thread()`). I can do the same for other `as_*_thread` methods as well (in this PR or another one) if people feel like this approach.

as_Worker_thread => WorkerThread::cast could pretty easily be added to this change.  There's only one call.

There are lots of as_Java_thread calls.  Maybe check with runtime team how they feel about the inconsistency vs code churn.

**Comment by iklam** [comment]
> > I encountered some circular dependencies after moving the definition to `thread.inline.hpp`, so I replaced it with a static `cast` method, as Kim instructed from an offline discussion. Note now there's a discrepancy: `CompilerThread::cast(t)` vs `t->as_Worker_thread()` (or `t->as_Java_thread()`). I can do the same for other `as_*_thread` methods as well (in this PR or another one) if people feel like this approach.
> 
> as_Worker_thread => WorkerThread::cast could pretty easily be added to this change. There's only one call.
> 
> There are lots of as_Java_thread calls. Maybe check with runtime team how they feel about the inconsistency vs code churn.

I prefer `JavaThread::cast()` over `Thread::as_Java_thread()`. The former style is consistent with other functions such as `InstanceKlass:cast()`.

As far as code churn, [JDK-8252685](https://bugs.openjdk.java.net/browse/JDK-8252685) already removed a bunch of `as_Java_thread()` calls (we have about 110 of them now, vs about 180 in JDK 16), so we may as well fix the rest for a more consistent style, which is also better with header dependencies. @dholmes-ora what do you think? In any case, we should do that in a different PR.

**Review by iklam — APPROVED** [review]
This looks good to me. If we decide to use Compiler::cast, the RFE title should be updated before integration.

**Review by dholmes-ora — APPROVED** [review]
The change to CompilerThread::cast doesn't seem necessary to address any circular header dependencies - what am I missing?

I personally prefer the as_X() style for thread casts (though X is spelt incorrectly in most places), but `cast` is a more traditional name and used for other type hierarchies in the VM. So we can change this for other thread types in a separate RFE.

As Ioi notes, this RFE needs a new synopsis: "Adopt cast notation for CompilerThread conversions" or something like that.

Thanks,
David

**Comment by dholmes-ora** [comment]
> The change to CompilerThread::cast doesn't seem necessary to address any circular header dependencies - what am I missing?

Doh! I was missing the fact the cast method goes in compilerThread.hpp not thread.hpp.

**Comment by albertnetymk** [comment]
> as_Worker_thread => WorkerThread::cast could pretty easily be added to this change. There's only one call.

I will do that in another PR to keep this one more focused.

> As Ioi notes, this RFE needs a new synopsis: "Adopt cast notation for CompilerThread conversions" or something like that.

Updated.

Thank you all for the suggestions and reviews.

/integrate

**Pull request closed without merging**